porch/controllers/downstreampackages/pkg/controllers/downstreampackage/downstreampackage_controller.go

```diff
@@ -0,0 +1,366 @@
+			pr.Spec.PackageName != downstream.Package {
+			continue
+		}
+		// check that the downstream package was created by this DownstreamPackage object
+		owned := false
+
+		// TODO: Implement the "AdoptionPolicy" field, which allows the user to decide if
+		//  the controller should adopt existing package revisions or ignore them. For now,
+		//  we just ignore them.
+		for _, owner := range pr.OwnerReferences {
+			if owner.UID == dp.UID {
+				owned = true
+			}
+		}
+		if !owned {
+			// this downstream package doesn't belong to us
+			continue
+		}
+
+		// Check if this PR is a draft. We should only have one draft created by this controller at a time,
+		// so we can just return it.
+		if pr.Spec.Lifecycle != porchapi.PackageRevisionLifecyclePublished {
+			return &pr
+		} else {
+			latestPublished, latestVersion = compare(&pr, latestPublished, latestVersion)
+		}
+	}
+
+	return latestPublished
+}
+
+func compare(pr, latestPublished *porchapi.PackageRevision, latestVersion string) (*porchapi.PackageRevision, string) {
+	switch cmp := semver.Compare(pr.Spec.Revision, latestVersion); {
+	case cmp == 0:
+		// Same revision.
+	case cmp < 0:
+		// current < latest; no change
+	case cmp > 0:
+		// current > latest; update latest
+		latestVersion = pr.Spec.Revision
+		latestPublished = pr.DeepCopy()
+	}
+	return latestPublished, latestVersion
+}
+
+// determine if the downstream PR needs to be updated
+func (r *DownstreamPackageReconciler) isUpToDate(dp *api.DownstreamPackage, downstream *porchapi.PackageRevision) bool {
+	upstreamLock := downstream.Status.UpstreamLock
+	lastIndex := strings.LastIndex(upstreamLock.Git.Ref, "/")
+	if strings.HasPrefix(upstreamLock.Git.Ref, "drafts") {
+		// the current upstream is a draft, so it needs to be updated.
+		return false
+	}
+	currentUpstreamRevision := upstreamLock.Git.Ref[lastIndex+1:]
+	return currentUpstreamRevision == dp.Spec.Upstream.Revision
+}
+
+func (r *DownstreamPackageReconciler) copyPublished(ctx context.Context,
+	source *porchapi.PackageRevision,
+	dp *api.DownstreamPackage) (*porchapi.PackageRevision, error) {
+	tr := true
+	newPR := &porchapi.PackageRevision{
+		TypeMeta: metav1.TypeMeta{
+			Kind:       "PackageRevision",
+			APIVersion: porchapi.SchemeGroupVersion.Identifier(),
+		},
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: source.Namespace,
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion:         dp.APIVersion,
+				Kind:               dp.Kind,
+				Name:               dp.Name,
+				UID:                dp.UID,
+				Controller:         &tr,
+				BlockOwnerDeletion: nil,
+			}},
+		},
+		Spec: source.Spec,
+	}
+
+	newPR.Spec.Revision = ""
+	newPR.Spec.WorkspaceName = newWorkspaceName(newPR.Spec.WorkspaceName)
+	newPR.Spec.Lifecycle = porchapi.PackageRevisionLifecycleDraft
+
+	if err := r.Client.Create(ctx, newPR); err != nil {
+		return nil, err
+	}
+
+	// get the newly created package revision back
+	var prList porchapi.PackageRevisionList
```

Comment:
The `newPR` variable should be updated with the latest state from the APIServer as part of the call to `Create`. so I don't think this is needed.

Reply:
done